### PR TITLE
cardano-cli: Add friendly printing of transactions (envelopes) with signatures

### PR DESCRIPTION
**Pull request by cblp**

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 1]
```diff
@@ -19,20 +19,44 @@ import qualified Data.ByteString.Char8 as BSC
...
+import qualified Data.Yaml.Pretty as Yaml
 
 import           Cardano.Api as Api
-import           Cardano.Api.Shelley (Address (ShelleyAddress), StakeAddress (..))
+import           Cardano.Api.Byron (KeyWitness (ByronKeyWitness))
+import           Cardano.Api.Shelley (Address (ShelleyAddress),
+                   KeyWitness (ShelleyBootstrapWitness, ShelleyKeyWitness), StakeAddress (..))
 import           Cardano.Ledger.Crypto (Crypto)
 import qualified Cardano.Ledger.Shelley.API as Shelley
+-- import           Cardano.Ledger.Shelley.TxBody (WitVKey (..))
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:31]: Left over comment?

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 2]
```diff
@@ -19,20 +19,44 @@ import qualified Data.ByteString.Char8 as BSC
...
 
+yamlConfig :: Yaml.Config
+yamlConfig = Yaml.defConfig & setConfCompare compare
+
+friendlyTxBS :: CardanoEra era -> Tx era -> ByteString
+friendlyTxBS era = Yaml.encodePretty yamlConfig . object . friendlyTx era
+
+friendlyTx :: CardanoEra era -> Tx era -> [Aeson.Pair]
+friendlyTx era (Tx body witnesses) =
+  ("witnesses" .= map friendlyKeyWitness witnesses) : friendlyTxBody era body
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:43]: We need to also render required signers.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 3]
```diff
@@ -85,8 +86,16 @@ friendlyTxBody
     , "update proposal" .= friendlyUpdateProposal txUpdateProposal
     , "validity range" .= friendlyValidityRange era txValidityRange
     , "withdrawals" .= friendlyWithdrawals txWithdrawals
+    , "witnesses-extra (payment key hashes needed for scripts)" .=
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:89]: this should be "required-signers"

**Review by Jimbo4350 — COMMENTED** [review]
Almost there

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 4]
```diff
@@ -19,20 +19,44 @@ import qualified Data.ByteString.Char8 as BSC
...
+
+friendlyTx :: CardanoEra era -> Tx era -> [Aeson.Pair]
+friendlyTx era (Tx body witnesses) =
+  ("witnesses" .= map friendlyKeyWitness witnesses) : friendlyTxBody era body
+
+friendlyKeyWitness :: KeyWitness era -> Aeson.Value
+friendlyKeyWitness =
+  object
+  . \case
+      ByronKeyWitness txInWitness -> ["input witness" .= textShow txInWitness]
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:48]: done

**Comment by cblp** [comment]
bors merge

**Pull request merged**